alpa/collective/collective_group/xla_nccl_collective_group.py

```diff
@@ -0,0 +1,393 @@
+                        f"somewhere else.")
+
+    @staticmethod
+    def generate_nccl_uid():
+        group_uid = xla_nccl_util.get_nccl_unique_id()
+        return group_uid
+
+    @staticmethod
+    def _generate_nccl_uid(key):
+        """Generate an NCCL unique ID for initializing communicators.
+
+        The method will also create a KV store using Ray named actor and store
+        the NCCLUniqueID in the store. The store needs to be garbage collected
+        when destroying the collective group.
+
+        Args:
+            key (str): the key for storage of NCCLUniqueID.
+
+        Returns:
+            NCCLUniqueID (str): NCCL unique ID.
+        """
+        group_uid = xla_nccl_util.get_nccl_unique_id()
+        store_name = get_store_name(key)
+        # Avoid a potential circular dependency in ray/actor.py
+        from alpa.collective.util import NCCLUniqueIDStore  # pylint: disable=import-outside-toplevel
+        store = NCCLUniqueIDStore.options(
+            name=store_name, lifetime="detached").remote(store_name)
+        ray.get([store.set_id.remote(group_uid)])
+        return group_uid
+
+    def create_p2p_communicator(self,
+                                my_gpu_idx: int,
+                                peer_rank: int,
+                                peer_gpu_idx: int,
+                                nccl_uid: str = None):
+        """A public method to create p2p communicators
+
+        Args:
+            my_gpu_idx (int): the gpu index on self rank.
+            peer_rank (int): the rank of the peer process.
+            peer_gpu_idx (int): the index of the gpu on the peer process.
+            nccl_uid (str, optional): optionally to provide the NCCLUniqueID in
+                advance.
+
+        Returns:
+            None
+        """
+        comm_key = _get_comm_key_send_recv(self.rank, my_gpu_idx, peer_rank,
+                                           peer_gpu_idx)
+        self._get_nccl_p2p_communicator(comm_key, my_gpu_idx, peer_rank,
+                                        peer_gpu_idx, nccl_uid)
+
+    def allreduce(self, tensors, allreduce_options=AllReduceOptions()):
+        raise NotImplementedError()
+
+    def barrier(self, barrier_options=BarrierOptions()):
+        raise NotImplementedError()
+
+    def reduce(self, tensors, reduce_options=ReduceOptions()):
+        raise NotImplementedError()
+
+    def allgather(self,
+                  tensor_lists,
+                  tensors,
+                  allgather_options=AllGatherOptions()):
+        raise NotImplementedError()
+
+    def broadcast(self, tensors, broadcast_options=BroadcastOptions()):
+        raise NotImplementedError()
+
+    def reducescatter(self,
+                      tensors,
+                      tensor_lists,
+                      reducescatter_options=ReduceScatterOptions()):
+        raise NotImplementedError()
+
+
+def _get_comm_key_send_recv(my_rank, my_gpu_idx, peer_rank, peer_gpu_idx):
```

Comment:
This is the same as `_get_comm_key_send_recv` in `nccl_collective_group`

Reply:
Thx! I will delete all cupy nccl in another PR when nccl XLA works for everyone.